packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1473,6 +1485,115 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
+    }
+    final Offset adjustedOffset = SelectionUtils.adjustDragOffset(
+      _rect,
+      localPosition,
+      direction: paragraph.textDirection,
+    );
+
+    final TextPosition position = paragraph.getPositionForOffset(adjustedOffset);
+    // Check if the original local position is within the rect, if it is not then
+    // we do not need to look up the word boundary for that position. This is to
+    // maintain a selectables selection collapsed at 0 when the local position is
+    // not located inside its rect.
+    final ({TextPosition wordStart, TextPosition wordEnd})? wordBoundary = !_rect.contains(localPosition) ? null : _getWordBoundaryAtPosition(position);
+    TextPosition? targetPosition;
+    if (wordBoundary != null) {
+      assert(wordBoundary.wordStart.offset >= range.start && wordBoundary.wordEnd.offset <= range.end);
+      if (_selectableContainsOriginWord && existingSelectionStart != null && existingSelectionEnd != null) {
+        final TextPosition edgeToKeepInPlace = isEnd ? existingSelectionStart : existingSelectionEnd;
+        final TextPosition currentMovingEdge = isEnd ? existingSelectionEnd : existingSelectionStart;
+        final bool shouldSwapEdgesWhenSelectionIsNormalized = isEnd ? position.offset < existingSelectionStart.offset : position.offset > existingSelectionEnd.offset;
+        final bool shouldSwapEdgesWhenSelectionNotNormalized = isEnd ? position.offset > existingSelectionStart.offset : position.offset < existingSelectionEnd.offset;
+        final bool shouldSwapEdges = existingSelectionStart.offset < existingSelectionEnd.offset ? shouldSwapEdgesWhenSelectionIsNormalized : shouldSwapEdgesWhenSelectionNotNormalized;
+        if (shouldSwapEdges) {
+          if (position.offset < edgeToKeepInPlace.offset) {
+            targetPosition = wordBoundary.wordStart;
+          } else {
+            targetPosition = wordBoundary.wordEnd;
+          }
+          final ({TextPosition wordStart, TextPosition wordEnd}) localWordBoundary = _getWordBoundaryAtPosition(edgeToKeepInPlace);
+          assert(localWordBoundary.wordStart.offset >= range.start && localWordBoundary.wordEnd.offset <= range.end);
+          _setSelectionPosition(edgeToKeepInPlace.offset == localWordBoundary.wordStart.offset ? localWordBoundary.wordEnd : localWordBoundary.wordStart, isEnd: !isEnd);
+        } else {
+          if (position.offset < edgeToKeepInPlace.offset) {
+            targetPosition = wordBoundary.wordStart;
+          } else if (position.offset > edgeToKeepInPlace.offset) {
+            targetPosition = wordBoundary.wordEnd;
+          } else {
+            targetPosition = currentMovingEdge;
+          }
+        }
+      } else {
+        final TextPosition? possibleExistingEdge = existingSelectionStart ?? existingSelectionEnd;
+        if (possibleExistingEdge != null && (possibleExistingEdge == existingSelectionStart && isEnd || possibleExistingEdge == existingSelectionEnd && !isEnd)) {
```

Comment:
```suggestion
        final TextPosition? possibleExistingEdge = isEnd ? existingSelectionStart : existingSelectionEnd;
        if (possibleExistingEdge != null) {
```

Reply:
As long as the non-moving edge is null you have to go to the else case no?

Reply:
also this may be easier if we separate out the isEnd = true/false into two separate functions